File: src/main/java/com/github/m50d/margatroid/parser/Parser.java
package com.github.m50d.margatroid.parser;

import java.util.Arrays;
import java.util.Iterator;
import java.util.Stack;
import java.util.function.Function;
import java.util.stream.Stream;

import com.github.m50d.margatroid.model.ast.Grouped;
import com.github.m50d.margatroid.model.ast.Literal;
import com.github.m50d.margatroid.model.ast.Node;
import com.github.m50d.margatroid.model.ast.Quoted;

public class Parser {
	Stream<Node> parse(Iterator<String> input) {
		Stream.Builder<Node> accumulator = Stream.builder();
		final Stack<Function<Stream<Node>, Stream.Builder<Node>>> stack = new Stack<>();
		while (input.hasNext()) {
			String next = input.next().intern();
			switch (next) {
			case "]":
			case "}":
				accumulator = stack.pop().apply(accumulator.build());
				break;
			case "[":
				Stream.Builder<Node> accumulator2 = accumulator;
				stack.push(sn -> accumulator2.add(new Grouped(sn)));
				accumulator = Stream.builder();
				break;
			case "{":
				Stream.Builder<Node> accumulator3 = accumulator;
				stack.push(sn -> accumulator3.add(new Quoted(sn)));
				accumulator = Stream.builder();
				break;
			default:
				accumulator.add(new Literal(next));
			}
		}
		assert stack.isEmpty();
		return accumulator.build();
	}

	public Stream<Node> parse(String input) {
		return parse(Arrays.asList(input.split(" ")).iterator());
	}
}

File: src/main/scala/com/github/m50d/margatroid/parser/Parser.scala
package com.github.m50d.margatroid.parser

import com.github.m50d.margatroid.model._
import scala.annotation.tailrec

class Parser {
  @tailrec
  final def parse(input: List[String], stack: List[IndexedSeq[AstNode] => IndexedSeq[AstNode]] = List.empty, accumulator: IndexedSeq[AstNode] = IndexedSeq.empty): IndexedSeq[AstNode] =
    input match {
      case Nil =>
        assert(stack isEmpty)
        accumulator
      case ("]" | "}") :: tail =>
        parse(tail, stack.tail, stack.head(accumulator))
      case "[" :: tail =>
        parse(tail, { ns: IndexedSeq[AstNode] => accumulator :+ Grouped(ns) } :: stack)
      case "{" :: tail =>
        parse(tail, { ns: IndexedSeq[AstNode] => accumulator :+ Quoted(ns) } :: stack)
      case sref :: tail if sref.startsWith("$") =>
        parse(tail, stack, accumulator :+ Reference(sref stripPrefix "$"))
      case literal :: tail =>
        parse(tail, stack, accumulator :+ Literal(literal))
    }

  def parse(input: String): IndexedSeq[AstNode] = parse(input.split(" ").toList)
}